src/mercury_engine_data_structures/formats/bmsld.py

```diff
@@ -34,125 +45,171 @@
-            unk13=StrId,
-            unk14=Hex(Int32ul),
+            "name" / StrId,
+            "unk01" / StrId,
+            "unk02" / Hex(Int32ul),
+            "unk03" / Hex(Int32ul),
+            "unk04" / Hex(Int32ul),
+            "unk05" / Hex(Int32ul),
+            "unk06" / Hex(Int32ul),
+            "unk07" / Hex(Int32ul),
+            "unk08" / Hex(Int32ul),
+            "position" / CVector3D,
+            "type" / StrId,
+            "unk14" / Hex(Int32ul),
         )
     ),
-    actors=make_dict(ProperActor)[18],
-    sub_areas=make_vector(
+    # layers for actors
+    "actor_layers" / make_dict(ProperActor)[18],
+    # collision_cameras and groups
+    "sub_areas"
+    / make_vector(
         Struct(
-            name=StrId,
-            names=make_vector(StrId),
+            "name" / StrId,
+            "objects" / make_vector(StrId),
         )
     ),
-    rest=construct.GreedyBytes,
+    # only used in s000_mainmenu, s010_cockpit, s020_credits
+    "extra_data" / construct.Optional(make_vector(ExtraActors)),
+    construct.Terminated,
 ).compile()
 
 
+class ActorLayer(IntEnum):
+    TRIGGER = 0
+    ENV_TRIGGER = 2
+    SPAWNGROUP = 3
+    SPAWNPOINT = 4
+    STARTPOINT = 5
+    PASSIVE = 9
+    PLATFORM = 10
+    DOOR = 15
+    CHOZO_SEAL = 16
+    HIDDEN_POWERUP = 17
+
+
+class BmsldActor:
+    def __init__(self, raw: Container) -> None:
+        self._raw = raw
+
+    @property
+    def actor_type(self) -> str:
+        return self._raw.type
+
+    @actor_type.setter
+    def actor_type(self, value: str) -> None:
+        self._raw.type = value
+
+    @property
+    def position(self) -> Vec3:
+        return Vec3(self._raw.position)
+
+    @position.setter
+    def position(self, value: Vec3) -> None:
+        self._raw.position = value.raw
+
+    @property
+    def rotation(self) -> Vec3:
+        return Vec3(self._raw.rotation)
+
+    @rotation.setter
+    def rotation(self, value: Vec3) -> None:
+        self._raw.rotation = value.raw
+
+    def get_component_function(self, component_idx: int = 0) -> Container:
+        return ComponentFunction(self._raw.component_functions[component_idx])
+
+
+ArgumentType = int | float | str | bool
+
+ARGUMENT_TYPES = {
+    "s": str,
+    "f": float,
+    "b": bool,
+    "i": int,
+}
+
+
+class ComponentFunction:
+    def __init__(self, raw: Container) -> None:
+        self._raw = raw
+
+    def set_argument(self, argument_idx: int, value: ArgumentType) -> None:
+        argument = self._raw.arguments[argument_idx]
+        expected_type = ARGUMENT_TYPES[argument.type]
+        if not isinstance(value, expected_type):
+            raise TypeError(f"Invalid argument type: expected {expected_type}, got {type(value)} ({value})")
+        argument.value = value
+
+
 class Bmsld(BaseResource):
     @classmethod
     def construct_class(cls, target_game: Game) -> Construct:
         return BMSLD
 
     def all_actors(self) -> Iterator[tuple[int, str, construct.Container]]:
```

Comment:
this should now return `Iterator[tuple[ActorLayer, str, BmsldActor]]` (you'll have to modify the function slightly to make it truth these)